rbdmap: replace rbdmap with systemd unit generator

huww98 opened the pull request:
Replace original rbdmap bash script with rbdmap-generator, which is a system unit generator that generates one rbdmap@.service for each RBD image on boot or `systemctl daemon-reload`.

This generator is written in Python3, and its behavior closely follows systemd-cryptsetup-generator(8). It is designed to be fully compatible with the original /etc/ceph/rbdmap config file.

Apart from original functions, this generator supports several new features:

* Support adding systemd unit options (e.g. x-systemd.requires=) from /etc/ceph/rbdmap config file. The design closely follows
  systemd-fstab-generator(8)

* Support noauto, nofail options. They have the same semantic as in systemd-cryptsetup-generator(8)

* Previous "id=admin,options='lock_on_read,queue_depth=1024'" can now be written as "id=admin,lock_on_read,queue_depth=1024", which is more nature.

We have some additional advantages compared with the original script:

* Other units may acquire dependency to a specific RBD device, may through the rbdmap@.service units, or the standard .device and blockdev@.target units.

* We do not have to deal with mount and umount in the script. The mount unit created by systemd-fstab-generator(8) will automatically pull in and wait until RBD device mapping finished.

* Furthermore, we may support on-demand RBD mapping. For example, when working with automount, the mount unit triggered by access to the mount point will in turn trigger the RBD device mapping automatically.

However, there are still some breaking changes:

* Manually invoke `rbdmap` command is not supported. They may be replaced with `systemctl start|stop rbdmap@*.service`. `rbdmap unmap-all` has no replacement.

* There is no easy way to override the config file path.

Hooks in /etc/ceph/rbd.d/ are still supported but should be considered deprecated. Administrators should use systemd unit file to deploy hooks.

Signed-off-by: 胡玮文 <huww98@outlook.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented:
@wjwithagen hi Willem, does the FreeBSD port use/support rbdmap?

tchaikov commented on ceph.spec.in:
```diff
@@ -1609,7 +1610,7 @@ exit 0
 %{_mandir}/man8/radosgw-admin.8*
 %{_mandir}/man8/rbd.8*
-%{_mandir}/man8/rbdmap.8*
+%{_mandir}/man8/rbdmap-generator.8*
```
might want to move this line to line 1627 -- right before `%{_unitdir}/rbdmap.target`. so it is closer to `rbdmap.target`.

huww98 commented on doc/man/8/rbdmap-generator.rst:
```diff
@@ -0,0 +1,140 @@
+Supported Options in Config File
+--------------------------------
+
+.. option:: conf=, cluster=, id=, name=, mon_host=, keyfile=, keyring=, device-type=, read-only, exclusive, quiesce=, quiesce-hook=
```
This line is too long and formatted badly on read-the-docs. Do you have any idea how can I fix this?

https://ceph--40844.org.readthedocs.build/en/40844/man/8/rbdmap-generator/#cmdoption-rbdmap-generator-arg-conf

tchaikov replied:
could just warp this line, like
```rst
.. option:: conf=, cluster=, id=, name=, mon_host=, keyfile=, keyring=,
   device-type=, read-only, exclusive, quiesce=, quiesce-hook=
```

huww98 replied:
Unfortunately, this makes the command-line man page ugly
![图片](https://user-images.githubusercontent.com/15710934/114701006-d74ea000-9d54-11eb-9636-004baf0adc76.png)

tchaikov replied:
indeed. how about using multiple `.. option::` directives?

tchaikov commented on doc/man/8/rbdmap-generator.rst:
```diff
@@ -0,0 +1,140 @@
+
+.. option:: conf=, cluster=, id=, name=, mon_host=, keyfile=, keyring=, device-type=, read-only, exclusive, quiesce=, quiesce-hook=
+
+   These options are passed to :doc:`rbd <rbd>`\(8) command directly as
```
```suggestion
   These options are passed to :program:`rbd`\(8) command directly as
```

as `rbd` is also defined as under the "program" domain, and can be referenced by `:program:` directly.

tchaikov commented on doc/man/8/rbdmap-generator.rst:
```diff
@@ -0,0 +1,140 @@
+
+.. option:: noauto, auto
+
+   With noauto, the device will not be added as a dependency for rbdmap.target.
```
might want to quote `noauto` with backquotes like
```rst
``noauto``
```

tchaikov commented on doc/man/8/rbdmap-generator.rst:
```diff
@@ -0,0 +1,140 @@
+   With noauto, the device will not be added as a dependency for rbdmap.target.
+   This means that it will not be automatically mapped on boot, unless something
+   else pulls it in. In particular, if the device is used for a mount point,
+   it'll be mapped automatically during boot, unless the mount point itself is
```
```suggestion
   it is mapped automatically during boot, unless the mount point itself is
```

huww98 replied:
This paragraph is mostly copied from https://www.freedesktop.org/software/systemd/man/crypttab.html#noauto. And I think the grammar is OK here.

tchaikov replied:
i don't think copying from an existing document could prove the correctness or readability of the resulting document. but i agree the grammar here is correct. what i am not sure is that "will" should be generally used in this context.

@zdover23 what do you think?

huww98 commented:
@wjwithagen `rbdmap` and `rbdmap-generator` now both parse `/etc/ceph/rbdmap` file, but I've added some new features that would break `rbdmap` once used. So I removed the `rbdmap` script to indicate this is an upgrade, and they may not be used interchangeably in the future.

But I think keeping the `rbdmap` script is also OK. We could state the above point in the man page. And do not distribute `rbdmap` on Linux platform by default. And I agree it is useful for other platforms.

wjwithagen commented:
@huww98 
Perhaps document the expected structure in /etc/ceph/rbdmap by yhe script.
And I could imagine running this script on FreeBSD on hypervisors to actually load images into the system with `rbd-ggate`
Something I never considered doing automated yet, but is of course needed

trociny commented:
I am just wondering, while we are here, couldn't be made more generic to also support other device types (nbd), as it is described in the feature request [1]?

Note, right now `rbd map|unmap` commands are just aliases to more generic `rbd device map|unmap --device-type kernel`. It would be nice if we could use this more generic variant, specifying the device-type (kernel,nbd) somewhere in the config.

[1] https://tracker.ceph.com/issues/20762

huww98 commented:
@trociny 
> I am just wondering, while we are here, couldn't be made more generic to also support other device types (nbd), as it is described in the feature request [1]?

I've just tested that it just works when you add `device-type=nbd` to the options list. `rbd map` also works with other device types.

In my case:
```
rbd_mixed/ImageNet id=admin,read-only,nofail,noauto,device-type=nbd
```

tchaikov commented on debian/ceph-common.install:
```diff
@@ -5,7 +5,9 @@ etc/bash_completion.d/rados
-lib/systemd/system/rbdmap.service
+lib/systemd/system/rbdmap-pre.target
+lib/systemd/system/rbdmap.target
+lib/systemd/system-generators/rbdmap-generator
```
i'd suggest update the CMakeLists.txt to install this script into `lib/systemd/system-generators/`. or install this file manually in `debian/rules`. otherwise it cannot be found by `dh_install`. see

```
dh_install: warning: ceph-common missing files: lib/systemd/system-generators/rbdmap-generator
	install -d debian/ceph-common//etc/bash_completion.d
	cp --reflink=auto -a debian/tmp/etc/bash_completion.d/ceph debian/tmp/etc/bash_completion.d/rados debian/tmp/etc/bash_completion.d/radosgw-admin debian/tmp/etc/bash_completion.d/rbd debian/ceph-common//etc/bash_completion.d/
	install -d debian/ceph-common//lib/systemd/system
	cp --reflink=auto -a debian/tmp/lib/systemd/system/ceph.target debian/tmp/lib/systemd/system/rbdmap-pre.target debian/tmp/lib/systemd/system/rbdmap.target debian/ceph-common//lib/systemd/system/
dh_install: error: missing files, aborting
```

https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2021-04-14-1504/a7e62fb9adf7839c7f9da5348f92794de16b35e6/default/265590/

huww98 replied:
Strange. I have tried to install it here: https://github.com/ceph/ceph/blob/c3e0ef1fecd448d9c56db3846c9aed0d8022849f/src/CMakeLists.txt#L863-L865

I will have a look at it.

tchaikov replied:
@huww98 ahh, i see. there are discrepancies between these two paths

- `libexec/systemd/system-generators/rbdmap-generator`
- `lib/systemd/system-generators/rbdmap-generator`

the former is the install path specified by CMake, the latter is specified for dh_install. instead of installing the generator script into `${CMAKE_INSTALL_LIBEXECDIR}/systemd/system-generators`, i think we should install it into `${CMAKE_INSTALL_LIBDIR}/systemd/system-generators`. you could query the `systemd_system_generator_dir` using `pkg-config`, like:

```cmake
find_package(PkgConfig QUIET REQUIRED)
pkg_check_modules(PKG_systemd QUIET systemd)
pkg_get_variable(systemd_system_generator_dir "systemd" "systemd_system_generator_dir")
if(NOT systemd_system_generator_dir)
  set(systemd_system_generator_dir "${CMAKE_INSTALL_LIBDIR}/systemd/system-generators")
end()
```

please note, i'd suggest avoid using `CMAKE` prefix for our own variables.

huww98 replied:
For now, I added this line to be consistent with other systemd things: https://github.com/ceph/ceph/blob/b7e57a079ac0ba387b8c2ec7ad93df37c02fd90d/debian/rules#L34

Maybe introduce `pkg_get_variable` in another PR?

idryomov reviewed (changes requested):
I left some comments but haven't looked at the new script itself yet.

In the cover letter you say

> It is designed to be fully compatible with the original /etc/ceph/rbdmap config file.

but then a few lines below

> Previous "id=admin,options='lock_on_read,queue_depth=1024'" can now be written as "id=admin,lock_on_read,queue_depth=1024", which is more nature.

Can the new script handle the old `options=...` syntax?

In a later comment you say

> I've added some new features that would break rbdmap once used.

Can you enumerate what those are, precisely?

idryomov commented:
> I've just tested that it just works when you add `device-type=nbd` to the options list. `rbd map` also works with other device types.
> 
> In my case:
> 
> ```
> rbd_mixed/ImageNet id=admin,read-only,nofail,noauto,device-type=nbd
> ```

I wonder we should try to make `device-type` more prominent.  Perhaps move it into its own column?  Just a thought, not sure if it's actually worth it.

tchaikov commented on debian/rules:
```diff
@@ -31,6 +31,7 @@ extraopts += -DCMAKE_INSTALL_LIBDIR=/usr/lib
 extraopts += -DCMAKE_INSTALL_LIBEXECDIR=/usr/libexec
 extraopts += -DCMAKE_INSTALL_SYSCONFDIR=/etc
 extraopts += -DCMAKE_INSTALL_SYSTEMD_SERVICEDIR=/lib/systemd/system
+extraopts += -DCMAKE_INSTALL_SYSTEMD_GENERATORDIR=/lib/systemd/system-generators
```
@huww98 could you please reconsider my suggestion at https://github.com/ceph/ceph/pull/40844/commits/b7e57a079ac0ba387b8c2ec7ad93df37c02fd90d#r613158301 ?

tchaikov commented on doc/start/quick-rbd.rst:
```diff
@@ -66,4 +66,4 @@ See `block devices`_ for additional details.
 .. _FAQ: http://wiki.ceph.com/How_Can_I_Give_Ceph_a_Try
 .. _OS Recommendations: ../os-recommendations
-.. _rbdmap manpage: ../../man/8/rbdmap
+.. _rbdmap-generator manpage: ../../man/8/rbdmap-generator
```
use `:program:` if you please. the relative path is fragile.

huww98 replied:
I changed this to ".. _rbdmap-generator manpage: :program:\`rbdmap-generator\`", but the link is broken. Can you suggest how exactly to use `:program:` in the link?

huww98 replied:
I've read some docs about sphinx, it seems `:program:` cannot be used to generate links. Now I use `:doc:` with absolute path. At least we can get a warning if the link breaks in the future.

tchaikov replied:
sorry, i was wrong. `:program:` is not supposed to be used here. yeah, `:doc:` is more appropriate.

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -11,6 +11,9 @@ set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
   set(CMAKE_INSTALL_RPATH "${CEPH_INSTALL_FULL_PKGLIBDIR}")
 endif()
+# for rbdmap-generator
+set(CMAKE_INSTALL_SYSTEMD_GENERATORDIR "${CMAKE_INSTALL_LIBEXECDIR}/systemd/system-generators"
```
again, please don't use CMAKE for our own variables.

tchaikov replied:
and `CMAKE_INSTALL_LIBEXECDIR` is proven to be wrong.

huww98 replied:
Just want to be consistent with https://github.com/ceph/ceph/blob/b7e57a079ac0ba387b8c2ec7ad93df37c02fd90d/systemd/CMakeLists.txt#L1-L2

If you don't like that, I will fix the above code first. But I would like to first investigate it. So maybe I will update after some days.

tchaikov commented on src/CMakeLists.txt:
```diff
@@ -856,8 +859,10 @@ if(WITH_RBD)
-    ${CMAKE_SOURCE_DIR}/src/rbdmap
     DESTINATION ${CMAKE_INSTALL_BINDIR})
+  install(PROGRAMS
+    ${CMAKE_SOURCE_DIR}/src/rbdmap-generator
```
nit, 
```cmake
install(PROGRAMS rbdmap-generator
  DESTINATION ${CMAKE_INSTALL_SYSTEMD_GENERATORDIR})
```
would suffice.

huww98 commented:
@idryomov
> Can the new script handle the old `options=...` syntax?

Yes, it can.

> In a later comment you say
> 
> > I've added some new features that would break rbdmap once used.
> 
> Can you enumerate what those are, precisely?

Basically everything listed in "new features" from cover-letter: `x-systemd` things, `noauto`, `nofail`, and passing options without write `options='a,b'`

trociny commented:
> I wonder we should try to make `device-type` more prominent. Perhaps move it into its own column? Just a thought, not sure if it's actually worth it.

:+1: for this

And I don't like very much we rely on `rbd map --device-type` working here. It works just by an accident. When `rdb device ...` commands were introduce the idea was that we would leave the old `rbd map|unmap|showmapped`  as aliases for the krbd usage only, just not to break existing scripts and because many users used to them. `--device-type` works there just because it was difficult to disable it. So I am not very happy we "legalize" this usage here.

Though it is just a point to consider. Actually I was glad to learn it already worked in some way for nbd.

huww98 commented:
@idryomov
> I wonder we should try to make `device-type` more prominent. Perhaps move it into its own column? Just a thought, not sure if it's actually worth it.

I think we should not introduce a new column. If we do, putting that column just after image-spec seems the most natural, but that would break the capability with the old rbdmap file. However, we may need some special treatment on `device-type`. e.g. Type=forking seems more suitable for nbd, while Type=oneshot should be used for krbd.

idryomov commented:
I don't see anything wrong with making it a third column.

huww98 commented:
Intuitively, /etc/fstab put type before options. Logically, some options are specific to certain device type, options depend on device type, so options should come after device type.

huww98 commented:
Hi all, I'm trying to get `--pid-file` work for `rbd-nbd` so that systemd can reliably determine the main process. However, this config is explicitly disabled here: https://github.com/ceph/ceph/blob/b7e57a079ac0ba387b8c2ec7ad93df37c02fd90d/src/tools/rbd_nbd/rbd-nbd.cc#L1588

This is initially introduced in ce7c1520a8019f011fa34dd898af317f78974577 to fix https://tracker.ceph.com/issues/16921. Basically, it fixes 2 clients using the same pid file. However, I scanned the commit history, and found no clue that we have ever defined a default value for `pid_file`. From [this comment](https://tracker.ceph.com/issues/16921#note-5), I assume the test case set up a `pid_file` in the global section of config file? So maybe we should fix the test case instead of disabling pid file for rbd-nbd?

Can anyone confirm that it is safe to remove this line? Or by other means I can enable pid file for rbd-nbd?

huww98 commented:
In this version, I have:
* fixed the systemd file install path as suggested by @tchaikov
* bring back old `rbdmap` script @wjwithagen
* use `rbd device map` to explicit support nbd

All discussion above as been addressed except:
* https://github.com/ceph/ceph/pull/40844#discussion_r613002159 read-the-docs formatting, minor
* https://github.com/ceph/ceph/pull/40844#discussion_r613012936 use "will" or not
* https://github.com/ceph/ceph/pull/40844#issuecomment-820168209 `rbd-nbd --pid-file` not working, maybe deferred to another PR.

Thank you for review this.

trociny commented:
> This is initially introduced in [ce7c152](https://github.com/ceph/ceph/commit/ce7c1520a8019f011fa34dd898af317f78974577) to fix https://tracker.ceph.com/issues/16921. Basically, it fixes 2 clients using the same pid file. However, I scanned the commit history, and found no clue that we have ever defined a default value for `pid_file`. From [this comment](https://tracker.ceph.com/issues/16921#note-5), I assume the test case set up a `pid_file` in the global section of config file? So maybe we should fix the test case instead of disabling pid file for rbd-nbd?

I think the problem was that when it was specified via config, e.g. something like this:
```
pid file = /var/run/ceph/$name.pid
```
it caused problems when mapping more than one image. And to have it useful in configs we would need to provide some useful placeholders likep `$pool` and `$image` so it would not conflict for different images mapped and still would be easy to see where to look for the pid file for a particular map.

And as it was rather common (at leas for our tests) to have `pid_file` set in the config in such a way that it would cause conflicts, it was decided to just disable it.

I am recalling we had discussion with Jason about it, but I don't remember what PR it was.

> Can anyone confirm that it is safe to remove this line? Or by other means I can enable pid file for rbd-nbd?

I don't see much problems with it but it might break some tests that will need update.

Taking that I don't see much use to be able to specify it via config (at least until some useful placeholders are provided), then may be allow to specify it via command line only? Because this is how you are going to use it.

Another option, couldn't you use `rbd-nbd list-mapped` output when looking for pid? Probably not, but just to make sure.

huww98 commented:
> Taking that I don't see much use to be able to specify it via config (at least until some useful placeholders are provided), then may be allow to specify it via command line only? Because this is how you are going to use it.

Yes, I want to specify it via command line. But how can I only allow command line but not config file? I think command-line option also updates `g_ceph_context->_conf`.

> Another option, couldn't you use `rbd-nbd list-mapped` output when looking for pid? Probably not, but just to make sure.

Not easily, I need to specify `PIDFile=` in systemd unit file. I can write a script to parse `list-mapped` output then generate a pid file. But I think this is an overkill. also missing `PIDFile=` seems not too bad. It also works. Just `PIDFile=` is recommended in systemd.service(5)

trociny commented:
> Yes, I want to specify it via command line. But how can I only allow command line but not config file? I think command-line option also updates `g_ceph_context->_conf`.

I suppose in `parse_args`, after calling `config.parse_config_files` (and may be `config.parse_env`, because I don't think it is a good idea to use it from env) you could reset the current value (`config.set_val("pid_file", "")`), and then safely remove `g_ceph_context->_conf.set_val_or_die("pid_file", "")` in `do_map`.

trociny commented:
> config.set_val("pid_file", "")

Ah sorry, it would not work. I forgot that `parse_args` result is not used for context initialization.

trociny commented:
> > config.set_val("pid_file", "")
> 
> Ah sorry, it would not work. I forgot that `parse_args` result is not used for context initialization.

A little different version. Add `pid_file` to `struct Config`. Do the same thing as suggested above in `parse_argse`, but after calling `config.parse_argv()` do `cfg->pid_file = config.get_val<std::string>("pid_file");` And then use this result in `do_map`: 
```
g_ceph_context->_conf.set_val_or_die("pid_file", cfg->pid_file);
```

Or just parse manually `args` in `do_map` with `ceph_argparse_witharg(..., "pid_file", ...)` before context initialization and use the result in `set_val_or_die`.

huww98 commented:
@trociny Thank you for such a detailed explanation. But I would like to not include this change in this PR. I think this is a hack and may cause confusion to future readers or users. And it needs more discussion.

Is it possible to update the tests like:
```
[global]
pid file = /var/run/ceph/$name.pid
[client]
pid file =
```
so that the `pid_file` from conf only apply to daemons but not clients?

trociny commented:
> @trociny Thank you for such a detailed explanation. But I would like to not include this change in this PR. I think this is a hack and may cause confusion to future readers or users. And it needs more discussion.

The argument parsing in the rbd-nbd is already hackish, mostly due to difficulty to combine the ceph options with "native" options in a nice way. I don't think yet one "hack" will make it worse, especially if we already use similar approach for some special options, like rbd default pool or data pool.

> Is it possible to update the config file in the tests like:
> 
> ```
> [global]
> pid file = /var/run/ceph/$name.pid
> [client]
> pid file =
> ```
> 
> so that the `pid_file` from conf only apply to daemons but not clients?

It is possible, the questions are what exactly configs we have to update and who will do this. If you are going to change the behavior you also have to provide the update for our tests in the same PR (I may help with testing this on teuthology though). Another problem is that it may be configured this way in real user environments and I don't think the users would be happy if the rbd-nbd started to mysteriously fail on start after upgrade.

So I'd rather we leave the old behavior (at leas for now, you said `PIDFile` is desirable but not necessary) and probably address it in the future, or make it use the command line option only.

idryomov commented on src/rbdmap-generator:
```diff
@@ -0,0 +1,282 @@
+        # service_type = 'Type=forking\nPIDFile=%t/%n.pid'
+        # rbd_map_args.append('pid-file=%t/%n.pid')
+        service_type = 'Type=forking'
+    elif device_type not in ('krbd', None):
```
Can you elaborate on what exactly do we lose by not having `PIDFile` (and hence not bothering with `--pid-file`)?  `rbd-nbd` is not a traditional service daemon after all, so I wonder what is the reasoning behind `Type=forking` instead of `Type=exec`?

huww98 replied:
`Type=exec` is new in systemd 240, and is not available to Ubuntu 18.04. And systemd will not monitor the forked `rbd-nbd` process.

`forking` because systemd will invoke `rbd`, which will invoke `rbd-nbd`. Then `rbd` process will exit after the RBD device is successfully mapped. After that, systemd will consider the `rbdmap@.servce` unit active, and will monitor the `rbd-nbd`, change the unit state accordingly when `rbd-nbd` exit. Then units depending on it may react to the state change. `PIDFile=` is used to tell systemd which process it should monitor (the MainPID). And if it is not provided, systemd will guess one.

idryomov replied:
Sorry, I meant `oneshot` -- the one that is used for `krbd`.

What I'm asking is if we even want systemd to monitor `rbd-nbd`.  Consider `krbd` case: there is nothing to monitor, if someone manually unmaps the image that has been mapped by the generated unit, systemd wouldn't know.  Is there a reason not to follow the same pattern for `rbd-nbd`?

huww98 replied:
That's true, and `oneshot` will work. But in the nbd case, we have a process to monitor, and systemd can monitor it. Why not do it? And I suppose someone may query systemd for the MainPID in the future.

In the krbd case, I also considered adding a `BindsTo=dev-rbd-xxx-xxx.device` to the generated unit file. So when the device is manually unmapped, we can deactivate `rbdmap@.service` unit. But `crypttab` seems does not have this.

huww98 replied:
One advantage of deactivating the unit properly is that systemd will try to map the RBD device again when the unit is pulled in again.

idryomov replied:
We need to be consistent between `krbd` and `nbd`.  There is no daemon or PID in the krbd case.

The `BindTo` thing sounds like something that might work the same way for both `krbd` and `nbd`, but given that `cryptsetup` doesn't do it, we probably shouldn't either.  Let's keep it simple for now.

huww98 replied:
IMO, there are some differences between `krbd` and `nbd`, and they are not entirely consistent naturally. `nbd` process can be killed (OOM, etc.), or crash (because of a bug, network issue, etc.). And administrators may decide to react when this does happen (e.g. by overriding the `RestartSec=`/`OnFailure=`). With `forking`, we definitely give local administrators (like me) more choices.

But if you insist, I will change it back to `simple`.

And just noticed, nbd will not create a symlink under `/dev/rbd/`, will it? Then although we can map the device, we cannot get mount in `/etc/fstab` to work, because we don't have a stable device path. Maybe we should delay the support for nbd.

idryomov replied:
They are not naturally consistent, but for the purposes of the generated units which are really about calling `rbd device map` and not much else, I think it makes sense to have them behave the same.

Yes, no symlink would be created for `nbd`.  As far as the generator goes, I think there should be no special cases for `nbd` in the generator script.  So whether it is delayed or not, the code should be more or less the same.

idryomov commented:
Following up on making `device-type` more prominent, perhaps we should move to a new config file?  `/etc/ceph/rbdtab`, similar to `/etc/crypttab`.  If `/etc/ceph/rbdtab` doesn't exist, fall back to `/etc/ceph/rbdmap` expecting the old syntax and don't allow the old `options=` construct in `/etc/ceph/rbdtab`.  The new format could be

```
<image spec>  <generic options>  <device type>  <map options>  <unmap options>
```

`<generic options>` (e.g. `id=admin`) are to be appended directly
`<map options>` are to be appended as `--options <map options>`
`<unmap options>` same as `<map options>` but for `rbd device unmap`

`x-systemd.*`, `noauto` and `nofail`, etc could go into `<generic options>` or have its own column.

What do you guys think?

huww98 commented:
@idryomov In that case, maybe we can only process the `/etc/ceph/rbdtab` with the new generator. and leave the `/etc/ceph/rbdmap` file to the old rbdmap bash script. Then we distribute both and let the administrator choose which one to use.

idryomov commented:
I think that would cause unnecessary confusion.  Which one should I use, which one am I using, etc?  Distributing only the new generator on Linux and leaving the old bash script in git for FreeBSD is the way to go IMO.

trociny commented:
> What do you guys think?

:+1: Though I would like ` <device type> ` be the first column or at least second just after the image spec.

> leaving the old bash script in git for FreeBSD

I don't think FreeBSD uses rbdmap in any way.

idryomov commented:
Making it the first column sounds good.

wjwithagen commented:
On 19-4-2021 19:12, Mykola Golub wrote:
> I don't think FreeBSD uses rbdmap in any way.
>
>
Not at the moment.

--WjW

tchaikov commented:
> On 19-4-2021 19:12, Mykola Golub wrote: I don't think FreeBSD uses rbdmap in any way.
> Not at the moment.
> […](#)

if that's the case. i'd suggest just drop it.

wjwithagen commented:
> 
> 
> > On 19-4-2021 19:12, Mykola Golub wrote: I don't think FreeBSD uses rbdmap in any way.
> > Not at the moment.
> > […](#)
> 
> if that's the case. i'd suggest just drop it.

I'm fine with that, since we do not have an attic to store old but usefull stuff in.
On the other hand is it something that will be useful and as such would it be nice 
if the new implementation to manage this process is "compatible" with systems 
that do not run systemd.

huww98 commented:
In this version:
* above mentioned rbdtab config file is implemented
* revert to use `Type=oneshot` for nbd device (no special case)
* rbdmap bash script is dropped again.

Thank you again for your review.

huww98 commented:
jenkins test make check

tchaikov commented:
> > > On 19-4-2021 19:12, Mykola Golub wrote: I don't think FreeBSD uses rbdmap in any way.
> > > Not at the moment.
> > > […](#)
> > 
> > 
> > if that's the case. i'd suggest just drop it.
> 
> I'm fine with that, since we do not have an attic to store old but usefull stuff in.

i think since we are using version control system, there is no need to preserve the maybe-useful-in-future stuff.

huww98 commented:
Could you review this again? Thanks.

tchaikov commented:
@idryomov @trociny  ping?

idryomov commented:
@huww98 This PR is on my list, I'll try to get to it next week.  Sorry for the delay!

trociny commented on src/rbdmap-generator:
```diff
@@ -0,0 +1,299 @@
+    }
+    systemd_unit_options = empty_systemd_unit_options()
+
+    device_type, rbd_spec, *cells = line.split(maxsplit=4)
```
May be not use this `maxsplit=4` limitation, just in case we want to add a new field in the future and still have the old scripts to work (ignoring it)?

idryomov commented on src/rbdmap-generator:
```diff
@@ -0,0 +1,273 @@
+            'rbd image or snapshot specification {} cannot starts or ends with /'
+            .format(spec))
+    if '/' not in spec:
+        spec = 'rbd/' + spec
```
I agree.  Let's be explicit and make it mandatory.

huww98 replied:
This is not easy. The default pool is 'rbd', this is already documented [here](https://docs.ceph.com/en/latest/man/8/rbd/#image-snap-group-and-journal-specs), so I think this is reliable. And we cannot only let the `rbd device map` command determine the default pool, because we need to know the pool name to determine the device path (`/dev/rbd/pool/image`).

So:
1. Make the pool name mandatory. But it is not mandatory for `rbd` command. They are inconsistent.
2. Rely on the already documented default pool name `rbd`.

Given the above, do you still prefer 1? I personally prefer 2 to keep things consistent.

trociny replied:
The default pool is not 'rbd' but what is set in `rbd_default_pool` configuration parameter. And it can be set/overridden either in global (mon centralized db) config, or in ceph.conf or in a command line argument or in `CEPH_ARGS` environmental variable.

So if you really want to support the default pool in your scripts, you would have to call e.g. `ceph config get client.$id rbd_default_pool` to get the name, instead of just using `rbd`.

Personally I would rather we didn't have this default pool at all from the start, because it makes more problems than goods. And it looks useful only for CLI to save users more typing. So let's not propagate it into the map file. I don't see any problem that pool is not mandatory for the CLI but will be mandatory for the rbdtab file.

trociny replied:
Ah, and `ceph config get client.$id rbd_default_pool` would not be enough. I though it would apply the local file override but it only looks the setting in the centralized db. So it is much tricker to get properly the default pool.

huww98 replied:
OK, thanks for the explaination. I will make it mandatory. This [rbd(8)](https://github.com/ceph/ceph/blob/master/doc/man/8/rbd.rst#image-snap-group-and-journal-specs) makes me think it is just hard-coded as "rbd". Maybe we should update it.

huww98 commented on src/rbdmap-generator:
```diff
@@ -0,0 +1,299 @@
+            working = ''
+        else:
+            working += c
+    yield working
```
I would like to, but cannot find a better solution. The proposed code will parse `id=admin,keyring=/etc/ceph/ceph.client.admin.keyring,options='lock_on_read,queue_depth=1024'` (from rbdmap man page) as
```python
[
    'id=admin',
    'keyring=/etc/ceph/ceph.client.admin.keyring',
    "options='lock_on_read",
    'queue_depth=1024'
]
```
, while the expected result is
```python
[
    'id=admin',
    'keyring=/etc/ceph/ceph.client.admin.keyring',
    "options='lock_on_read,queue_depth=1024'"
]
```

trociny replied:
Ah, I see. I was thinking it was just needed to strip quotes from the option lists. Ack.

huww98 commented on doc/man/8/rbdmap.rst:
```diff
@@ -1,128 +0,0 @@
-:orphan:
```
But man page section 8 is for system administration commands. Since we have removed the `rbdmap` executable, we should also remove this page, or it may cause confusion. Maybe we can add another rbdmap man page to section 5 (File formats and conventions)?

trociny replied:
My suggestion is to add a note at the beginning of the manpage that the tool has been removed and the man is still here just to provide information about /etc/ceph/rbdmap format which is still supported for compatibility reasons although deprecated.

But I don't have a strong opinion here and will be actually fine with any solution.

huww98 commented on doc/man/8/rbdmap-generator.rst:
```diff
@@ -0,0 +1,153 @@
+
+    krbd bar1
+    krbd foopool/bar2    id=admin,noauto
+    krbd foopool/bar3    id=admin        lock_on_read,queue_depth=1024  force
```
Good point. I think maybe we can switch to a more complex config format? e.g. INI style like [systemd.netdev(5)](https://www.freedesktop.org/software/systemd/man/systemd.netdev.html)

`/etc/ceph/rbd/foopool-bar2.rbd`
```ini
[RBD]
Spec=foopool/bar2
Auto=no
Id=admin

[Mount]
LockOnRead=yes
QueueDepth=1024

[Umount]
force=yes
```

huww98 replied:
crypttab uses "none" or "-" as a placeholder for its third field, we can also follow it. So that we can write `krbd foopool/bar3 none lock_on_read`. What do you think?

trociny replied:
LGTM

huww98 commented:
I fixed a link in the doc in a2ccd9f, for other changes, please see git commit log. Thanks for your review.

trociny reviewed (approved):
LGTM (not being an expert in systemd though)

tchaikov commented:
@idryomov ping?

@huww98  needs rebase.

tchaikov commented:
@huww98 please drop the merge commit from this changeset. you might want to do the rebase locally instead of pushing the merge button.

huww98 commented:
@tchaikov Done. Dropped the merge commit, and rebased onto the latest master branch.

tchaikov commented:
jenkins test make check

huww98 commented:
Rebased and dropped the 2 already merged commits.

tchaikov commented:
@idryomov want to take a final look?

huww98 commented:
I would like to share our experience of using this on ~20 servers.

* We use automount to trigger the rbd device map on Ubuntu. `fstrim` would unintentionally trigger the map(util-linux/util-linux#1463), which seems fixed, but not backported to Ubuntu 20.04.
* The rbd device disappeared once (maybe due to a network issue), but the systemd unit remains active. Admin needs to restart the unit manually to recover.

tchaikov commented:
i am adding the needs-qa tag, so it can be picked up in a rados qa run.

idryomov commented:
unstale

yuriw commented:
@huww98 pls rebase

tchaikov commented:
> @huww98 pls rebase

@huww98 i rebase your change. but looks like your PR does not allow the edit of maintainers. so i cannot update it. could you rebase your change?

@idryomov ping?

huww98 commented:
Rebased. Thanks. @tchaikov

idryomov commented:
> @idryomov ping?

My apologies, I'm swamped with higher-priority work which is intended to be backported to squid right now.  I'll try to take a look at this as time allows in the coming weeks.  My main concern is that we don't have test coverage (neither for the old way nor for the new way of doing mappings), so this is something that would need to be play-tested manually.

> I would like to share our experience of using this on ~20 servers.
> 
>  * We use automount to trigger the rbd device map on Ubuntu. `fstrim` would unintentionally trigger the map([fstrim: ignore automount util-linux/util-linux#1463](https://github.com/util-linux/util-linux/issues/1463)), which seems fixed, but not backported to Ubuntu 20.04.
> 
>  * The rbd device disappeared once (maybe due to a network issue), but the systemd unit remains active. Admin needs to restart the unit manually to recover.

@huww98 Are you continuing to use this production?  Anything new to share?

huww98 commented:
We deploy this in our college. I have graduated now, so I'm not aware of the current status.
